src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -118,8 +223,137 @@ public Uri GetEndpointForClient(ConfigurationClient client)
             }
 
             ConfigurationClientWrapper currentClient = _clients.FirstOrDefault(c => c.Client == client);
-            
+
+            if (_dynamicClients != null && currentClient == null)
+            {
+                currentClient = _dynamicClients.FirstOrDefault(c => c.Client == client);
+            }
+
             return currentClient?.Endpoint;
         }
+
+        public void SetLogger(Logger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _logger = logger;
+        }
+
+        private async Task OptimisticRefreshFallbackClients(DateTimeOffset now, CancellationToken cancellationToken)
+        {
+            if (_replicaDiscoveryEnabled &&
+                now >= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval &&
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
```

Comment:
We seem to have lost the functionality of "add a replica on the fly to recover". In a scenario where all clients were failed over, we previously would have discovered new replica clients using the minimum refresh interval. Not the case with the new change.

I think we should be fine if we update the refresh to take into account the state where all replicas are failed over, and expedite the refresh in such a case.

```

                (_dynamicClients == null ||
                _dynamicClients.All(c => c.BackoffEndTime > now)
                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval)
                ```

Reply:
Will need a method name update with that change since it's no longer just optimistic refresh.

Reply:
I have some different idea, if a store don't have any replicas, this beget looking up SRV DNS every time while trying to get available clients. Even all dynamic clients are exhausted, it's also not necessary to lookup DNS every time while get available client.

Reply:
So I think we still need judge `now >= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval` to avoid too frequent DNS lookup 

```
               (now>= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval &&
                (_dynamicClients == null ||
                _dynamicClients.All(c => c.BackoffEndTime > now) ||
                now>= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
```